### PR TITLE
executor: use deep copy for maxMin4JSON

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve? 
fix https://github.com/pingcap/tidb/issues/15278

### What is changed and how it works?
change shallow copy to deep copy

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

use large dataset to reproduce this problem.

Code changes

 - No

Side effects

 - No

Related changes

 - Need to cherry-pick to the release branch

Release note

 - No

### CODE REVIEW

File: executor/aggfuncs/func_max_min.go [hunk 1]
```diff
@@ -635,7 +635,7 @@ func (e *maxMin4JSON) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup [
 		}
 		cmp := json.CompareBinary(input, p.val)
 		if e.isMax && cmp > 0 || !e.isMax && cmp < 0 {
-			p.val = input
+			p.val = input.Copy()
```
Comment (SunRunAway) [inline executor/aggfuncs/func_max_min.go]: Could you add some tests in `TestAggJSON`?

File: executor/aggfuncs/func_max_min.go [hunk 2]
```diff
@@ -652,7 +652,7 @@ func (e *maxMin4JSON) MergePartialResult(sctx sessionctx.Context, src, dst Parti
 	}
 	cmp := json.CompareBinary(p1.val, p2.val)
 	if e.isMax && cmp > 0 || !e.isMax && cmp < 0 {
-		p2.val = p1.val
+		p2.val = p1.val.Copy()
```
Comment (SunRunAway) [inline executor/aggfuncs/func_max_min.go]: It this not necessary?

Comment (fzhedu) [reply executor/aggfuncs/func_max_min.go]: recovered.

File: executor/aggfuncs/func_max_min.go (hunk 1 above)
Comment (fzhedu) [reply executor/aggfuncs/func_max_min.go]: added

**Comment by fzhedu** [comment]
> > solve the bug results from using shallow copy in maxMin4JSON
> 
> Could you create an issue describing the detail of the bug for this PR?

It is hard to reproduce this issue using a small data set.

**Comment by zz-jason** [comment]
> > > solve the bug results from using shallow copy in maxMin4JSON
> > 
> > 
> > Could you create an issue describing the detail of the bug for this PR?
> 
> It is hard to reproduce this issue using a small data set. So, I cannot.

1. you can create a large data set to reproduce the problem
2. you can use a small chunk size to reproduce the problem

File: executor/aggregate_test.go [hunk 3]
```diff
@@ -830,3 +830,16 @@ func (s *testSuiteAgg) TestIssue12759HashAggCalledByApply(c *C) {
 		tk.MustQuery(tt).Check(testkit.Rows(output[i]...))
 	}
 }
+func (s *testSuiteAgg) TestPR15242ShallowCopy(c *C) {
```
Comment (SunRunAway) [inline executor/aggregate_test.go]: Please add an empty line between two functions

Comment (fzhedu) [reply executor/aggregate_test.go]: ok

**Comment by fzhedu** [comment]
> > > > solve the bug results from using shallow copy in maxMin4JSON
> > > 
> > > 
> > > Could you create an issue describing the detail of the bug for this PR?
> > 
> > 
> > It is hard to reproduce this issue using a small data set. So, I cannot.
> 
> 1. you can create a large data set to reproduce the problem
> 2. you can use a small chunk size to reproduce the problem

Another important factor is to reduce the system stack size: `ulimit -s 33`

**Pull request merged**